modules/host/network.go

```diff
@@ -20,57 +21,116 @@ func (h *Host) initNetworking(address string) error {
+	return nil
 }
 
-// handleConn handles an incoming connection to the host, typically an RPC.
-func (h *Host) handleConn(conn net.Conn) {
-	defer conn.Close()
+// threadedHandleConn handles an incoming connection to the host, typically an
+// RPC. threadedHandleConn is responsible for releasing a readlock on
+// host.resourceLock when all communications have completed.
+func (h *Host) threadedHandleConn(conn net.Conn) {
+	// All threaded functions are called holding the close lock, and are
+	// expected to keep it until they no longer need access to the host's
+	// resources.
+	defer h.resourceLock.RUnlock()
+	// If the host resources are unavailable, return early.
+	if h.closed {
+		return
+	}
+
 	// Set an initial duration that is generous, but finite. RPCs can extend
-	// this if so desired.
-	conn.SetDeadline(time.Now().Add(5 * time.Minute))
+	// this if desired.
+	err := conn.SetDeadline(time.Now().Add(5 * time.Minute))
+	if err != nil {
+		h.log.Println("WARN: could not set deadline on connection:", err)
+		return
+	}
+	defer conn.Close()
 
+	// Read a specifier indicating which action is beeing called.
 	var id types.Specifier
 	if err := encoding.ReadObject(conn, &id, 16); err != nil {
+		atomic.AddUint64(&h.atomicMalformedCalls, 1)
+		h.log.Printf("WARN: incoming conn %v was malformed", conn.RemoteAddr())
 		return
 	}
-	var err error
+
 	switch id {
-	case modules.RPCSettings:
-		err = h.rpcSettings(conn)
-	case modules.RPCUpload:
-		err = h.rpcUpload(conn)
+	case modules.RPCDownload:
+		atomic.AddUint64(&h.atomicDownloadCalls, 1)
+		err = h.managedRPCDownload(conn)
 	case modules.RPCRenew:
-		err = h.rpcRenew(conn)
+		atomic.AddUint64(&h.atomicRenewCalls, 1)
+		err = h.managedRPCRenew(conn)
 	case modules.RPCRevise:
-		err = h.rpcRevise(conn)
-	case modules.RPCDownload:
-		err = h.rpcDownload(conn)
+		atomic.AddUint64(&h.atomicReviseCalls, 1)
+		err = h.managedRPCRevise(conn)
+	case modules.RPCSettings:
+		atomic.AddUint64(&h.atomicSettingsCalls, 1)
+		err = h.managedRPCSettings(conn)
+	case modules.RPCUpload:
+		atomic.AddUint64(&h.atomicUploadCalls, 1)
+		err = h.managedRPCUpload(conn)
 	default:
+		atomic.AddUint64(&h.atomicErroredCalls, 1)
 		h.log.Printf("WARN: incoming conn %v requested unknown RPC \"%v\"", conn.RemoteAddr(), id)
 		return
 	}
 	if err != nil {
+		atomic.AddUint64(&h.atomicErroredCalls, 1)
 		h.log.Printf("WARN: incoming RPC \"%v\" failed: %v", id, err)
 	}
 }
 
-// rpcSettings is an rpc that returns the host's settings.
-func (h *Host) rpcSettings(conn net.Conn) error {
+// listen listens for incoming RPCs and spawns an appropriate handler for each.
+func (h *Host) threadedListen() {
+	// All threaded are called holding a readlock, and must release the
+	// readlock upon terminating.
+	defer h.resourceLock.RUnlock()
+
+	// If the host has closed, some of the necessary resources will not be
+	// available, and therefore the host should not be receiving connections.
+	if h.closed {
+		return
+	}
```

Comment:
currently, yes. but if the host hits the bandwidth cap or something it might make sense to start toggling this. In the meantime, I'll turn this into a sanity check